sympy/solvers/solvers.py

```diff
@@ -1120,41 +1146,82 @@ def _has_piecewise(e):
     # try to get a solution
     ###########################################################################
     if bare_f:
-        solution = _solve(f[0], *symbols, **flags)
-        # solution is:
-        # dict for coeficient system with one solution
-        # list of values
-        # list of dicts
+        solution = None
+        # check for undetermined coefficients situation: `solve(a*x + 2*x + b - c, a, b)`
+        if len(symbols) != 1:
+            g = f[0]
+            free = g.free_symbols                        # {a, b, c, x}
+            ex = free - symset                           # {c, x}
+            if len(ex) != 1:
+                ind, dep = g.as_independent(*symbols)    # (2*x - c, a*x + b)
+                ex = ind.free_symbols & dep.free_symbols # {x, c} & {a, x, b} -> {x}
+            if len(ex) != 1:                             # e.g. (a+b)*x + b - c -> {c}, {a, b, x}
+                ex = dep.free_symbols - symset           # {x} = {a,b,x}-{a.b}
+            if len(ex) == 1:
+                ex = ex.pop()
+                try:
+                    # force dict return for solution
+                    missing = Dummy()
+                    was = {k: flags.pop(k, missing) for k in ('dict', 'set')}
+                    flags['dict'] = True
+                    # get solution which (because it will call solve with
+                    # flags) will be simplified and checked
+                    soln = solve_undetermined_coeffs(g, symbols, ex, **flags)
+                    # restore keys that weren't missing after
+                    # removing added 'dict'
+                    flags.pop('dict', None)
+                    for k in was:
+                        if was[k] != missing:
+                            flags[k] = was[k]
```

Comment:
One of significant causes of spaghetti code in this function is all of the manipulation of flags. Why manipulate flags here instead of creating a new dict to be passed to `solve_undetermined_coeffs`?

It takes a lot of though to understand what is going on here because the code is so confusing and I'm certain it can be simplified. For example a pattern that I see you using often is something like
```python
was = {k: flags.pop(k, missing) for k in ('dict', 'set')}
```
Here we are using a relatively complicated construct (a dict comprehension) just to loop over two items. It's usually not worth having a loop over two items rather than just a little bit of duplication e.g.:
```python
was = {'dict': flags.pop('dict', None), 'set': flags.pop('set', None)}
```
In any case I think it is better not to use `pop` here because we shouldn't be modifying flags at all.

The situation here is that we want to pass the flags through except we want to override the `set` and `dict` flags. The dict union operator from https://peps.python.org/pep-0584/ can do this nicely but requires Python 3.9 while SymPy still supports Python 3.8. In any case the PEP lists a number of alternative spellings.

The code highlighted above can just be:
```python
rflags = dict(flags, set=None, dict=True)
soln = solve_undetermined_coeffs(g, symbols, ex, **rflags)
```